SSA/Experimental/Bits/Fast/Reflect.lean

```diff
@@ -0,0 +1,671 @@
+    let mExample := f!"{w1} → {wExample1}" ++ f!"{w2} → {wExample2}"
+    throwError "found multiple bitvector widths in the target: {indentD mExample}"
+  else
+    -- we have exactly one width
+    let (w, wExample) := ws[0]
+    if !w.isFVar then
+      throwError "expcted width to be a free variable, found with '{w}' (for bitvector '{wExample}')"
+    else
+      -- invariant: w is known to be an fvar.
+      let wFv := w.fvarId!
+      -- We can now revert hypotheses that are of this bitwidth.
+      let g ← revertBvHyps g
+          
+      -- Next, after reverting, we have a goal which we want to reflect.
+      -- we convert this goal to NNF
+      let .some g ← NNF.runNNFSimpSet g
+        | logInfo m!"simp automatically closed goal."
+          return[]
+      logInfo m!"goal after NNF: {indentD g}"
+      -- finally, we perform reflection.
+      let result ← reflectPredicateAux ∅ (← g.getType)
+      let bvToIxMapVal ← result.bvToIxMap.toExpr w
+      let target := (mkAppN (mkConst ``Predicate.denote) #[result.e, w, bvToIxMapVal])
+      let g ← g.replaceTargetDefEq target
+      logInfo m!"goal after reflection: {indentD g}"
+      let (mapFv, g) ← generalizeMap g bvToIxMapVal; 
+      let (_, g) ← g.revert #[mapFv]
+      -- Apply Predicate.denote_of_eval_eq.
+      let [g] ← g.apply <| (mkConst ``Predicate.denote_of_eval_eq) 
+        | throwError m!"Failed to apply `Predicate.denote_of_eval_eq` on goal '{indentD g}'"
+      let [g] ← g.apply <| (mkConst ``of_decide_eq_true) 
+        | throwError m!"Failed to apply `of_decide_eq_true on goal '{indentD g}'"
+      let [g] ← g.apply <| (mkConst ``Lean.ofReduceBool) 
+        | throwError m!"Failed to apply `of_decide_eq_true on goal '{indentD g}'"
+      return [g]
+
+/--
+Given a goal state of the form:
+  ∀ (w : Nat)
+  ∀ (b₁ b₂ ... bₙ : BitVec w),
+  <proposition about bitvectors>.
+
+decide the property by reduction to finite automata.
+
+TODO(@bollu): Also decide properties about finite widths, by extending to the maximal width and clearing the high bits?
+-/
+elab "bv_reflect" : tactic => do
+  liftMetaTactic fun g => do
+    reflectUniversalWidthBVs g
+
+
+elab "bv_automata3" : tactic => do
+  liftMetaTactic fun g => do
+    reflectUniversalWidthBVs g 
+
+  match ← getUnsolvedGoals  with
+  | [] => return ()
+  -- | TODO: replace with ofReduceBool
+  | [_g] => evalDecideCore `bv_automata3 (cfg := { native := true : Parser.Tactic.DecideConfig})
+  | _gs => throwError "expected single goal after reflecting, found multiple goals. quitting"
+  
+/-- Can solve explicitly quantified expressions with intros. bv_automata3. -/
+theorem eq1 : ∀ (w : Nat) (a : BitVec w), a = a := by
+  intros 
+  bv_automata3
+#print eq1
+
+/-- Can solve implicitly quantified expressions by directly invoking bv_automata3. -/
+theorem eq2 (w : Nat) (a : BitVec w) : a = a := by
+  bv_automata3
+#print eq1
+
+
+open NNF in
+
+example (w : Nat) (a b : BitVec w) : a + b = b + a := by
```

Comment:
@tobiasgrosser I've left examples here, and I'm still adding to them at the moment. Feel invited to add new ones, which will help me flag down what isn't working right now!